contracts/balance/contract.go

```diff
@@ -284,6 +327,121 @@ func Burn(from interop.Hash160, amount int, txDetails []byte) {
 	runtime.Log("assets were burned")
 }
 
+// epochBillingStat is a copy of github.com/nspcc-dev/neofs-contract/contracts/container.EpochBillingStat
+// to prevent cross-contract imports that may fail due to internal `_deploy` calls.
+type epochBillingStat struct {
+	Account interop.Hash160
+
+	LatestContainerSize    int
+	LatestEpoch            int
+	LastUpdateTime         int
+	LatestEpochAverageSize int
+
+	PreviousContainerSize    int
+	PreviousEpoch            int
+	PreviousEpochAverageSize int
+}
+
+// PayForContainer distributes storage payments from container's owner account
+// to storage nodes that serve container's objects. Transaction must be
+// witnessed by the actual Alphabet multi-signature. Produces `ChangePaymentStatus`
+// notification on any payment status changes. If payment is successful, `Payment`
+// notification is thrown. If payment cannot be fully made, container is
+// registered as an unpaid one, see [IsUnpaid].
+func PayForContainer(cid interop.Hash256) bool {
+	if len(cid) != interop.Hash256Len {
+		panic("invalid container id")
+	}
+	common.CheckAlphabetWitness()
+
+	var (
+		ctx                   = storage.GetContext()
+		netmapContractAddr    = storage.Get(ctx, netmapContractKey).(interop.Hash160)
+		containerContractAddr = storage.Get(ctx, containerContractKey).(interop.Hash160)
+
+		containerUserNeoFS = contract.Call(containerContractAddr, "owner", contract.ReadOnly, cid).([]byte)
+		containerOwner     = common.WalletToScriptHash(containerUserNeoFS)
+		rate               = contract.Call(netmapContractAddr, "config", contract.ReadOnly, balanceconst.BasicIncomeRateKey).(int)
+		currEpoch          = contract.Call(netmapContractAddr, "epoch", contract.ReadOnly).(int)
+
+		transferredTotal int
+		paymentOK        = true
+	)
+
+	it := contract.Call(containerContractAddr, "iterateBillingStats", contract.ReadOnly, cid).(iterator.Iterator)
+	for iterator.Next(it) {
+		var (
+			r    = iterator.Value(it).(epochBillingStat)
+			size int
+		)
+
+		switch {
+		case r.LatestEpoch < currEpoch-1:
+			// no updates for more than 2 epochs, consider load be the same
+			size = r.LatestContainerSize
+		case r.LatestEpoch == currEpoch:
+			if r.PreviousEpoch == currEpoch-1 {
+				size = r.PreviousEpochAverageSize
+			} else {
+				size = r.PreviousContainerSize
+			}
+		case r.LatestEpoch == currEpoch-1:
+			size = r.LatestEpochAverageSize
+		default:
+			size = 0
+		}
+
+		payment := rate * size / gigabyte
+		if payment == 0 {
```

Comment:
Have you considered all rounding effects? https://github.com/nspcc-dev/neofs-node/issues/3357 point 6.

Reply:
not sure i understand you. owner need to pay `rate * size / gigabyte` to every node. what exactly are you speaking about?

Reply:
Consider a single 1 byte object. It'll never be paid for if you don't do a ceiling rounding to 10^-12. If size is not zero and if the rate is not zero (if it's zero there is nothing to do anyway).